src/phoenix/server/api/routers/v1/dataset_examples.py

```diff
@@ -0,0 +1,68 @@
+from sqlalchemy import func, select
+from starlette.responses import JSONResponse, Response
+from starlette.status import HTTP_404_NOT_FOUND
+from strawberry.relay import GlobalID
+
+from phoenix.db.models import DatasetExample, DatasetExampleRevision
+
+
+async def list_dataset_examples(request):
+    dataset_id = GlobalID.from_id(request.path_params["id"])
+    raw_version_id = request.query_params.get("version")
+    version_id = GlobalID.from_id(raw_version_id) if raw_version_id else None
+
+    async with request.app.state.db() as session:
+        if (type_name := dataset_id.type_name) != "Dataset":
+            return Response(
+                content=f"ID {dataset_id} refers to a f{type_name}", status_code=HTTP_404_NOT_FOUND
+            )
+
+        # Subquery to find the maximum created_at for each dataset_example_id
+        # timestamp tiebreaks are resolved by the largest id
+        subquery = select(
+            DatasetExampleRevision.dataset_example_id,
+            func.max(DatasetExampleRevision.created_at).label("most_recent"),
+            func.max(DatasetExampleRevision.id).label("max_id"),
+        ).group_by(DatasetExampleRevision.dataset_example_id)
+
+        # if a version_id is provided, filter the subquery to only include revisions from that
+        # version or earlier
+        if version_id is not None:
+            subquery = subquery.filter(
+                (DatasetExampleRevision.dataset_version_id <= int(version_id.node_id))
+            )
+
+        subquery = subquery.subquery()
+
+        # Query for the most recent example revisions that are not deleted
```

Comment:
Is it not safe to assume that row ids are monotonically increasing?

Reply:
it depends on what we want monotonicity to be a function of, if we're talking about as a function of insertion time, yeah, I think it's pretty reliable. But if we're talking about request time, under concurrent loads this may not be the case. I don't think it's worth getting too complex here even though I'm experimenting with a few possibly more robust approaches, and I definitely make the assumption that DB index is monotonic in the way "we want" at the end of the day